### PR TITLE
[WIP] Begin comment confirm page.

### PR DESCRIPTION
**Pull request by jmcarp**
* Submit comments via synchronous POST
* Add skeleton of comment confirm page
* Add comment confirm view

There's plenty left to do here, but the basics are working, and I think the approach should be clear enough to review if folks want to take a look.

### CODE REVIEW

**Comment by jmcarp** [comment]
Here's the current behavior, unstyled:

d7zmfqtuoj

To capture a discussion from slack: as of this proposal, when the confirmation page loads, we don't have access to the pdf download url or the tracking number. We're going to check periodically for both until they exist. The pdf url may show up sooner than the tracking number, although in the screen capture they appear at the same time.

@cmc333333 @vrajmohan: have time to review?

File: regulations/tasks.py [hunk 1]
```diff
@@ -24,55 +25,70 @@
...
 from django.template import loader
+from django.utils.crypto import get_random_string
 
 from regulations.models import FailedCommentSubmission
 logger = get_task_logger(__name__)
 
 
 @shared_task(bind=True)
-def submit_comment(self, body):
+def submit_comment(self, comments, form, meta):
```
Comment (vrajmohan) [inline regulations/tasks.py]: Would `additional_comment_fields` be a better name than `form`?
Would `signed_url` be a better name than `meta`?

File: regulations/tasks.py [hunk 2]
```diff
@@ -24,55 +25,70 @@
...
     '''
     Submit the comment to regulations.gov. If unsuccessful, retry the task.
     Number of retries and time between retries is managed by Celery settings.
     The main comment is converted to a PDF and added as an attachment; the
     'general_comment' field refers to this attachment.
 
-    :param body: dict of fields and values to be posted to regulations.gov
+    :param comments: List of sectional comments
+    :param form: Dict of form data
+    :param meta:
```
Comment (vrajmohan) [inline regulations/tasks.py]: Describe.

File: regulations/tasks.py [hunk 3]
```diff
@@ -24,55 +25,70 @@
...
+def submit_comment(self, comments, form, meta):
     '''
     Submit the comment to regulations.gov. If unsuccessful, retry the task.
     Number of retries and time between retries is managed by Celery settings.
     The main comment is converted to a PDF and added as an attachment; the
     'general_comment' field refers to this attachment.
 
-    :param body: dict of fields and values to be posted to regulations.gov
+    :param comments: List of sectional comments
+    :param form: Dict of form data
```
Comment (vrajmohan) [inline regulations/tasks.py]: Would it be better to replace "Dict of form data" with "Additional comment fields accepted by regulations.gov"?

File: regulations/templates/regulations/comment-confirm-chrome.html [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        <li>
+          Then EPA will post your comment to <a href="#" target="_blank">the docket folder for
+          this proposed rule</a> (with <a href="#" target="_blank">a few exceptions</a>). You
+          will be able to find your comment there or by searching
+          <a href="https://regulations.gov" target="_blank">Regulations.gov</a>
+          for your comment tracking number or keywords from your comment.
+        </li>
+      </ul>
+
+      <h4>EPA will review it & publish a final rule</h4>
```
Comment (vrajmohan) [inline regulations/templates/regulations/comment-confirm-chrome.html]: Minor - use &amp;

File: regulations/static/regulations/js/source/views/comment/comment-confirm-view.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    this.docId = this.$el.data('doc-id');
+    this.metaUrl = this.$el.data('meta-url');
+
+    this.listenTo(CommentEvents, 'read:proposal', this.handleRead);
+
+    if (this.metaUrl) {
+      this.poll(this.metaUrl);
+      comments.filter(this.docId).forEach(function(comment) {
+        comment.destroy();
+      });
```
Comment (cmc333333) [inline regulations/static/regulations/js/source/views/comment/comment-confirm-view.js]: Can we add a test for this, that comments disappear? It's a critical path

File: regulations/static/regulations/js/source/views/comment/comment-confirm-view.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  },
+
+  handleRead: function() {
+    var section = this.docId + '-preamble-' + this.docId + '-I';
+    var options = {id: section, section: section, mode: 'read'};
+
+    $('#content-body').removeClass('comment-review-wrapper');
+
+    MainEvents.trigger('section:open', section, options, 'preamble-section');
+  },
```
Comment (cmc333333) [inline regulations/static/regulations/js/source/views/comment/comment-confirm-view.js]: This feels a little nasty -- for example, I don't know that we can guarantee that section `I` will always be present. Why is this needed?

File: regulations/static/regulations/js/source/views/comment/comment-review-view.js [hunk 7]
```diff
@@ -67,12 +66,13 @@ var CommentReviewView = Backbone.View.extend({
     this.$content.html(html);
     this.findElms();
 
+    this.$form.find('[name="comments"]').val(JSON.stringify(commentData));
+
     this.preambleHeadView = new PreambleHeadView();
     CommentEvents.trigger('comment:writeTabOpen');
   },
 
   serialize: function() {
```
Comment (cmc333333) [inline regulations/static/regulations/js/source/views/comment/comment-review-view.js]: Given that this isn't used for general serialization anymore, should we rename it or inline it into `preview`?

File: regulations/static/regulations/js/source/views/comment/comment-confirm-view.js (hunk 6 above)
Comment (jmcarp) [inline regulations/static/regulations/js/source/views/comment/comment-confirm-view.js]: I don't disagree, but this is taken directly from `CommentReviewView`--@xtine may be able to say more about where it came from. We can have the template pass through the id of the first section in the toc instead--or just use a non-ajax link, since ajax navigation doesn't totally work with this view anyway.

File: regulations/static/regulations/js/source/views/comment/comment-confirm-view.js (hunk 5 above)
Comment (jmcarp) [inline regulations/static/regulations/js/source/views/comment/comment-confirm-view.js]: Prefer a unit test for the view or fleshing out the existing end-to-end comment submit test? Or both?

File: regulations/tasks.py [hunk 8]
```diff
@@ -24,55 +25,71 @@
...
 
 @shared_task
-def publish_tracking_number(response, key):
+def publish_tracking_number(response, metadata_url):
     s3 = make_s3_client()
-    body = {'trackingNumber': response['trackingNumber']}
+    body = {
+        'pdfUrl': response['pdfUrl'],
+        'trackingNumber': response['trackingNumber'],
+    }
```
Comment (cmc333333) [inline regulations/tasks.py]: Curious: is `response` just the return value from `submit_comment`? If so, would it make sense to skip the `body` here and just `json.dumps` the response?

File: regulations/tasks.py [hunk 9]
```diff
@@ -173,8 +222,7 @@ def build_multipart_encoded(body, comment_pdf, attachments):
     # Add other submitted fields
     fields.extend([
         (name, value)
-        for name, value in six.iteritems(body)
-        if name != 'assembled_comment'
+        for name, value in six.iteritems(form)
     ])
```
Comment (cmc333333) [inline regulations/tasks.py]: Can this be shortened to `fields.extend(form.items())` now?

File: regulations/tests/tasks_tests.py [hunk 10]
```diff
@@ -25,48 +26,69 @@ class TestSubmitComment(TestCase):
...
+        result = submit_comment(self.comments, self.form, self.meta)
 
-        self.assertEqual(result, expected_result)
+        self.assertEqual(
+            result,
+            {
+                'trackingNumber': 'some-tracking-number',
+                'pdfUrl': cache_pdf.return_value.url
+            },
+        )
```
Comment (cmc333333) [inline regulations/tests/tasks_tests.py]: Good! This test wasn't realistic before.

File: regulations/static/regulations/js/source/views/comment/comment-confirm-view.js (hunk 5 above)
Comment (cmc333333) [inline regulations/static/regulations/js/source/views/comment/comment-confirm-view.js]: I think an integration test would be good, but worry about how to mock that w/ s3. If it's a pain, a unittest around this specific logic will have to do.

File: regulations/static/regulations/js/source/views/comment/comment-confirm-view.js (hunk 6 above)
Comment (cmc333333) [inline regulations/static/regulations/js/source/views/comment/comment-confirm-view.js]: I'd vote for non-ajax link at this stage.

Comment (xtine) [inline regulations/static/regulations/js/source/views/comment/comment-confirm-view.js]: This definitely needs a change. It was temporarily like that to get read mode to work, I wasn't able at the time to figure out how to get to the main proposal. But now that metadata is a bit more clear to handle, this should be replaced.

File: regulations/views/comment.py [hunk 11]
```diff
@@ -70,62 +66,66 @@ def preview_comment(request):
...
-    )
-    chain.delay()
-    return JsonResponse({
-        'status': 'submitted',
-        'metadata_url': metadata_url,
-    })
+    def post(self, request, doc_number):
+        form = {
+            key: value for key, value in request.POST.items()
+            if key not in {'comments', 'csrfmiddlewaretoken'}
```
Comment (cmc333333) [inline regulations/views/comment.py]: I understand the desire to exclude `comments`, but let's avoid spelling out `csrfmiddlewaretoken` -- it'll be stripped in the sanitization step (as it would be if we used any other form framework). My concern is around readability.

File: regulations/views/comment.py [hunk 12]
```diff
@@ -70,62 +66,66 @@ def preview_comment(request):
...
+        }
+        comments = json.loads(request.POST.get('comments', '[]'))
+
+        context = common_context(doc_number)
+        context.update(generate_html_tree(context['preamble'], request,
+                                          id_prefix=[doc_number, 'preamble']))
+        context.update({'message': None, 'meta': None})
```
Comment (cmc333333) [inline regulations/views/comment.py]: `meta`'s not a safe keyword. It's also used to indicate the meta data about the proposal as a whole.

File: regulations/views/comment.py [hunk 13]
```diff
@@ -70,62 +66,66 @@ def preview_comment(request):
...
+
+        valid, context['message'] = self.validate(comments, form)
+
+        try:
+            context['meta'] = self.enqueue(comments, form)
+        except Exception as exc:
```
Comment (cmc333333) [inline regulations/views/comment.py]: Can this exception be more specific? What are we expecting will go wrong?

File: regulations/templates/regulations/comment-confirm-chrome.html [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+          EPA will process your comment, which may take a few days or weeks depending
+          on the number of comments received. In a few cases EPA may decline to evaluate
+          or post your comment, as listed in its
+          <a href="https://www.epa.gov/dockets/commenting-epa-dockets#rules" target="_blank">
+            rules and restrictions for comments
+          </a>.
+        </li>
+        <li>
+          In most cases, EPA will post your comment to
+          <a href="#" target="_blank">the docket folder for this proposed rule</a>.
```
Comment (cmc333333) [inline regulations/templates/regulations/comment-confirm-chrome.html]: Mind making an issue for filling this in?

File: regulations/templates/regulations/comment-review-chrome.html [hunk 15]
```diff
@@ -62,41 +62,48 @@ <h4 class="preview-button" disabled>Building your PDF...</h4>
...
+
+          <input type="checkbox" class="agree">
+          <label for="agree">
+            I read and understand the statement above.
+            <span class="required">(Required to submit.)</span>
+          </label>
+        </div>
+
+        <button class="submit-button" disabled>Submit your comment</button>
+        <div class="status"></div>
```
Comment (cmc333333) [inline regulations/templates/regulations/comment-review-chrome.html]: Is this div still needed?

File: regulations/views/comment.py (hunk 13 above)
Comment (jmcarp) [inline regulations/views/comment.py]: The way I've been thinking about this, there are three possible outcomes of calling this view: success, failure due to invalid data, and general failure due to our system. This block checks for the general failure case. That might happen if we're unable to connect to the broker, or get an error writing to the broker. I looked at this a bit yesterday, and some of those errors might be broker-specific, which makes them harder to enumerate here.

File: regulations/static/regulations/js/source/views/comment/comment-confirm-view.js (hunk 6 above)
Comment (cmc333333) [inline regulations/static/regulations/js/source/views/comment/comment-confirm-view.js]: :+1: made an issue to handle it separately: https://github.com/eregs/notice-and-comment/issues/232

File: regulations/templates/regulations/comment-review-chrome.html (hunk 15 above)
Comment (jmcarp) [inline regulations/templates/regulations/comment-review-chrome.html]: Nope, deleting.

File: regulations/views/comment.py (hunk 13 above)
Comment (cmc333333) [inline regulations/views/comment.py]: Roger -- would you mind adding a comment to the code explaining the same?

**Comment by jmcarp** [comment]
Pushed a bunch of miscellaneous updates, mostly making things more concise and renaming variables. I added unit tests around clearing comments when the confirm page loads--I think it's worth putting some thought into integration tests, but we have to deal with celery (relatively easy--we can configure celery to run everything synchronously), wkhtmltopdf, and s3, which adds up to kind of a mess.

**Comment by cmc333333** [comment]
Good stuff
